File: 0216-combination-sum-iii/0216-combination-sum-iii.py
class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        dp = [[[] for _ in range(0, n+1)] for _ in range(0, 10)]
        for i in range(0, 10):
            dp[i][0].append([])
        for cur_num in range(1, 10):
            for cur_target in range(1, n+1):
                dp[cur_num][cur_target] += dp[cur_num-1][cur_target]
                if cur_num <= cur_target:
                    for sub in dp[cur_num-1][cur_target-cur_num]:
                        if len(sub) < k:
                            dp[cur_num][cur_target].append(sub + [cur_num])
        return [x for x in dp[-1][-1] if len(x) == k]
        
        #
        
        res = []
        
        def backtrack(i: int, cur_list: list[int], n: int) -> None:
            if n == 0 and len(cur_list) == k:
                res.append([x for x in cur_list])
                return
            if 9 >= i <= n:
                backtrack(i+1, cur_list + [i], n-i)
                backtrack(i+1, cur_list, n)
        
        backtrack(1, [], n)
        return res

File: 0216-combination-sum-iii/0216-combination-sum-iii.ts
function combinationSum3(k: number, n: number): number[][] {
    const dp = Array(10).fill(0).map(() => Array(n+1).fill(0).map(() => Array(k+1).fill(0).map(() => [])))
    for (let i=0; i < 10; i++) {
        dp[i][0][0].push([])
    }
    for (let curNum=1; curNum < 10; curNum++) {
        for (let curTarget=1; curTarget <= n; curTarget++) {
            for (let i=1; i <= k; i++) {
                dp[curNum][curTarget][i] = clone(dp[curNum-1][curTarget][i])
                if (curNum <= curTarget) {
                    for (const sub of dp[curNum-1][curTarget-curNum][i-1]) {
                        dp[curNum][curTarget][i].push([...sub, curNum])
                    }
                }
            }
        }
    }
    return dp[9][n][k]
    
    function clone(a: number[][][][] | number[][][] | number[][] | number[] | number) {
        if (typeof a === "number") {
            return a
        } else {
            return a.map(clone)
        }
    }
};